[lit] Update internal shell lexer to handle LLDB persistent vars.

cmtice opened the pull request:
LLDB allows creation of 'persistent' variables, with names that start with '$'. The lit internal shell was escaping the '$', making it '\\$', in some CHECK lines, which causes an LLDB test, TestExprWithSideEffectOnConvenienceVar, to fail when using the lit internal shell.

Further explanation of the failing LLDB test: LLDB convenience variables start with '$'. The test passes several quoted commands that use and update convenience variables to lldb as arguments to be run in batch mode.  The tool that packages up the complete string and passes it to the lit internal shell lexer for lexing inserts a backslash in front of the '$' before passing the string in for lexing. The lexer was passing this change along, causing the tests to fail. 

This PR fixes the issue by having the lexer remove the newly added escape on the '$'.

cmtice committed:
[lit] Update internal shell lexer to handle LLDB persistent vars.

LLDB allows creation of 'persistent' variables, with names that
start with '$'. The lit internal shell was escaping the '$',
making it '\$', in some CHECK lines, which causes some LLDB tests
to fail when using the lit internal shell.

This PR fixes that by having the lexer remove the escape on the '$'.

arichardson commented:
Is it possible to include a test for this?

boomanaiden154 reviewed (commented):
This needs a test in `llvm/utils/lit/tests`.

Adding some information in the PR description on what LLDB tests are impacted by this would also be helpful. There are some existing issues related to enabling the lit internal shell in LLDB, and adding a link to those (particularly https://github.com/llvm/llvm-project/issues/102698, none of the others look relevant) wouldn't hurt.

cmtice committed:
Merge remote-tracking branch 'origin/main' into lit-shell-clang-extra

cmtice committed:
Add test for converting '\$' to '$' in the lexer.

cmtice committed:
update lexer test

cmtice commented:
> This needs a test in `llvm/utils/lit/tests`.
> 
> Adding some information in the PR description on what LLDB tests are impacted by this would also be helpful. There are some existing issues related to enabling the lit internal shell in LLDB, and adding a link to those (particularly #102698, none of the others look relevant) wouldn't hurt.

I updated an existing internal shell lexer test to test for this.  I also updated the PR description, giving a lot more detail.  Also mentioned this PR in issue #102698.

cmtice committed:
Fix format issue.

arichardson commented:
> Just making a note here in case anyone else comes across it:
> 
> Certain sequences after this change need to be double escaped if they are supposed to be passed to the command escaped, like the following from an lld test:
> 
> ```
> # RUN: awk '/^__OBJC_\\$_CATEGORY_MyBaseClass_\\$_Category01:/ { print; getline; sub(/^[ \t]*\.quad[ \t]+l_OBJC_CLASS_NAME_$/, "\t.quad\tL_OBJC_IMAGE_INFO+3"); print; next } { print }' merge_cat_minimal.s > merge_cat_minimal_bad_name.s
> ```

Is this behaviour different from POSIX sh? I think we should make sure that the internal shell has the same behaviour.

I think maybe the difference we have here is that the removal of the `\` should only happen for double-quoted strings but not single-quoted ones?

cmtice commented on llvm/utils/lit/tests/unit/ShUtil.py:
```diff
@@ -28,7 +28,8 @@ def test_quoting(self):
         self.assertEqual(self.lex(""" "" "" """), ["", ""])
         self.assertEqual(self.lex(""" a\\ b """, win32Escapes=True), ["a\\", "b"])
-
+        self.assertEqual(self.lex('"\\$y = 11"'), ["$y = 11"])
+        self.assertEqual(self.lex('"expr \\$y = 11"'), ["expr $y = 11"])
```
See https://github.com/llvm/llvm-project/pull/156742 for requested fix & test.
